File: Revision/STL/heaps.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // max heap
    priority_queue<int> q;

    q.push(20);
    q.push(2);
    q.push(5);
    q.push(50);

    cout<<q.top()<<endl; // 50
    cout<<q.size()<<endl; // 4
    cout<<q.empty()<<endl; //0
    
    while(!q.empty()){
        cout<<q.top()<<" ";
        q.pop();
    }
    cout<<endl;

    // min heap
    priority_queue<int,vector<int> , greater<int>> pq;

    pq.push(100);
    pq.push(10);
    pq.push(5);
    pq.push(20);

    cout<<pq.top()<<endl;
    cout<<pq.size()<<endl;
    cout<<pq.empty()<<endl;

    while(!pq.empty()){
        cout<<pq.top()<<" ";
        pq.pop();
    }
    return 0;
}

File: java/Heaps.java
import java.util.*;

public class Heaps{
    public static void main(String[] args){

        // create a priority queue
        // by default it is min Heap
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int[] arr = {10,20,30,11,22,4,5,6};

        // add elements into heap
        // add() --> TC = O(logn)
        // remove() --> TC = O(logn)
        // peek() --> Tc = O(1)

        // Tc = nlogn
        for(Integer val : arr){
            pq.add(val);
        }

        // print heap
        while(pq.size() > 0){
            System.out.println(pq.peek());
            pq.remove();
        }

        System.out.println("------------------------");

        // max Heap
        PriorityQueue<Integer> pq1 = new PriorityQueue<>(Collections.reverseOrder());

        for(Integer val : arr){
            pq1.add(val);
        }

        // print heap
        while(pq1.size() > 0){
            System.out.println(pq1.peek());
            pq1.remove();
        }
    }
}